s3/pkg/service/object.go

```diff
@@ -363,6 +364,255 @@ func (s *s3Service) GetObject(ctx context.Context, req *pb.GetObjectInput, strea
+	}
+
+	if in.CopyType == utils.MoveType_ChangeStorageTier {
+		log.Infof("chagne storage class of %s\n", targetObject.ObjectKey)
+		// just change storage tier
+		targetBucket = srcBucket
+		className, err := GetNameFromTier(in.TargetTier, srcBackend.Type)
+		if err != nil {
+			return ErrInternalError
+		}
+		err = srcSd.ChangeStorageClass(ctx, targetObject, &className)
+		if err != nil {
+			log.Errorf("change storage class of object[%s] failed, err:%v\n", targetObject.ObjectKey, err)
+			return err
+		}
+		// TODO: update storage class in meta
+		newObj := &meta.Object{Object: targetObject}
+		err = s.MetaStorage.UpdateObject(ctx, srcObject, newObj)
+	} else {
+		// need move data, get target location first
+		if in.CopyType == utils.MoveType_ChangeLocation {
+			targetBucket = srcBucket
+			targetObject.Location = in.TargetLocation
+			log.Infof("move %s cross backends, srcBackend=%s, targetBackend=%s, targetTier=%d\n",
+				srcObject.ObjectKey, srcObject.Location, targetObject.Location, targetObject.Tier)
+		} else { // MoveType_MoveCrossBuckets
+			log.Infof("move %s from bucket[%s] to bucket[%s]\n", targetObject.ObjectKey, srcObject.BucketName,
+				targetObject.BucketName)
+			targetBucket, err = s.MetaStorage.GetBucket(ctx, in.TargetBucket, true)
+			if err != nil {
+				log.Errorf("get bucket[%s] failed with err:%v\n", in.TargetBucket, err)
+				return err
+			}
+			targetObject.ObjectKey = in.TargetObject
+			targetObject.Location = targetBucket.DefaultLocation
+			targetObject.BucketName = targetBucket.Name
+			log.Infof("move %s cross buckets, targetBucket=%s, targetBackend=%s, targetTier=%d\n",
+				srcObject.ObjectKey, targetObject.BucketName, targetObject.Location, targetObject.Tier)
+		}
+
+		// get storage driver
+		targetBackend, err := getBackend(ctx, s.backendClient, targetObject.Location)
+		if err != nil {
+			log.Errorln("failed to get backend client with err:", err)
+			return err
+		}
+		targetSd, err = driver.CreateStorageDriver(targetBackend.Type, targetBackend)
+		if err != nil {
+			log.Errorln("failed to create storage. err:", err)
+			return err
+		}
+
+		// steps of moving object data: add target object for gc -> copy data -> update meta(remove target object from
+		// gc, update object to be the target one, and add source object for gc in a transaction) -> delete source
+		// object data from backend-> delete source object from gc. If crash happened, gc service will clean data.
+		// step 1: add new object for gc
+		newObj := &meta.Object{Object: targetObject}
+		err = s.MetaStorage.AddGcobjRecord(ctx, newObj)
+		if err != nil {
+			log.Errorf("failed to add gcobj record[%v], err:%v", newObj, err)
+			// if delete failed, no error return, because gc will clean it
+			ierr := s.MetaStorage.DeleteGcobjRecord(ctx, newObj)
+			log.Infof("delete source object from gc finished, err:", ierr)
+			return err
+		}
+		// step 2: copy data
+		err = s.copyData(ctx, srcSd, targetSd, srcObject.Object, targetObject)
```

Comment:
after the copy, it is better to check the etag(md5, if has) with the original one to make sure the data integrity.

Reply:
Thanks for your suggestion, it's fixed, please review again.